asf_heat_pump_suitability/analysis/hn_zones/comparison_of_hn_zones.py

```diff
@@ -0,0 +1,350 @@
+LSOA_SHP_PATH = "s3://asf-heat-pump-suitability/source_data/Lower_layer_Super_Output_Areas_2021_EW_BFE_V9_-9107090204806789093/LSOA_2021_EW_BFE_V9.shp"
+NESTA_HP_SUITABILITY_PARQUET_PATH = "s3://nesta-open-data/asf_heat_pump_suitability/2023Q4/20240925_2023_Q4_EPC_heat_pump_suitability_per_lsoa.parquet"
+LA_TO_ANALYSE = "Liverpool"
+
+
+def setup_logging_and_file_path(
+    output_dir: str, log_filename: str = "script_output.log", level: int = logging.INFO
+) -> None:
+    """
+    Set up logging configuration to log messages to both a file and the console.
+
+    Args:
+        output_dir (str): Path to the directory where the log file should be saved.
+        log_filename (str): Name of the log file. Defaults to "script_output.log".
+        level (int): Logging level, e.g., logging.INFO or logging.DEBUG. Defaults to logging.INFO.
+    """
+    # Ensure the output directory exists
+    os.makedirs(output_dir, exist_ok=True)
+
+    # Define the full path to the log file
+    log_file_path = os.path.join(output_dir, log_filename)
+
+    # Clear existing logging handlers if any
+    for handler in logging.root.handlers[:]:
+        logging.root.removeHandler(handler)
+
+    # Set up logging configuration
+    logging.basicConfig(
+        level=level,
+        format="%(levelname)s: %(message)s",
+        handlers=[
+            logging.FileHandler(log_file_path, mode="w"),  # Overwrite log file each run
+            logging.StreamHandler(),
+        ],
+    )
+    logging.info(f"Logging setup complete. Logs are saved to {log_file_path}.")
+
+
+def load_hn_geodata(
+    desnz_hn_gpkg_path: str, lsoa_shp_path: str
+) -> Tuple[gpd.GeoDataFrame, List[str]]:
+    """
+    Load the HN GeoPackage file and perform a spatial join with LSOA polygons to add a column of LSOA codes.
+
+    Args:
+        desnz_hn_gpkg_path (str): Path to the GeoPackage file with the HN zones.
+        lsoa_shp_path (str): Path to the LSOA shapefile.
+
+    Returns:
+        Tuple[gpd.GeoDataFrame, List[str]]:
+            - GeoDataFrame with LSOA codes added via spatial join.
+            - List of unique LSOA codes present in the GeoDataFrame.
+    """
+    # Load DESNZ HN polygons GeoPackage
+    desnz_hn_gdf = pyogrio.read_dataframe(
+        desnz_hn_gpkg_path, layer="heat-network-zone-map-Liverpool"
+    )
+
+    # Load the LSOA polygons file with 'LSOA21CD' codes
+    lsoa_gdf = gpd.read_file(lsoa_shp_path)
+
+    # Remove 'index_right' column if it exists
+    if "index_right" in desnz_hn_gdf.columns:
+        desnz_hn_gdf = desnz_hn_gdf.drop(columns=["index_right"])
+
+    # Ensure both GeoDataFrames use the same CRS
+    if desnz_hn_gdf.crs != desnz_hn_gdf.crs:
+        lsoa_gdf = lsoa_gdf.to_crs(desnz_hn_gdf.crs)
+
+    # Perform spatial join to match Liverpool polygons with LSOA codes
+    joined_gdf = gpd.sjoin(
+        desnz_hn_gdf,
+        lsoa_gdf[["geometry", "LSOA21CD"]],
+        how="left",
+        predicate="intersects",
+    )
```

Comment:
```suggestion
    joined_gdf = desnz_hn_gdf.sjoin(
        lsoa_gdf[["geometry", "LSOA21CD"]],
        how="left",
        predicate="intersects",
    )
```

This is buried in the `geopandas` documentation but it's suggested to use `sjoin` as a method rather than function: https://geopandas.org/en/stable/docs/user_guide/mergingdata.html#spatial-joins

> For historical reasons, both methods are also available as top-level functions [sjoin()](https://geopandas.org/en/stable/docs/reference/api/geopandas.sjoin.html#geopandas.sjoin) and [sjoin_nearest()](https://geopandas.org/en/stable/docs/reference/api/geopandas.sjoin_nearest.html#geopandas.sjoin_nearest). It is recommended to use methods as the functions may be deprecated in the future.

Reply:
Good spot!